### PR TITLE
Trusted Device Encryption

### PR DESCRIPTION
**Pull request by Hinton**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Implement trusted device encryption. Verified backwards compatibility with existing device keys from desktop.

## Before you submit

- Please add **unit tests** where it makes sense to do so

### CODE REVIEW

File: crates/bitwarden-crypto/src/keys/device_key.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+/// Device Key
+///
+/// Encrypts the DevicePrivateKey
+/// Allows the device to decrypt the UserKey, via the DevicePrivateKey.
+#[derive(Debug)]
+pub struct DeviceKey(SymmetricCryptoKey);
+
+#[derive(Debug)]
+pub struct CreateDeviceKey {
```
Comment (jlf0dev) [inline crates/bitwarden-crypto/src/keys/device_key.rs:14]: nit: Not sure of the naming scheme here but at a glance this tells me these are the inputs needed to create a Device Key. Maybe this should be `CreatedDeviceKey`?

**Review by jlf0dev — APPROVED** [review]
Looks good!

File: crates/bitwarden-crypto/src/keys/device_key.rs (hunk 1 above)
Comment (Hinton) [reply crates/bitwarden-crypto/src/keys/device_key.rs]: Good point, I've renamed it to TrustDeviceResponse which is inline with other structs. We'll want to revisit this in the future.

File: crates/bitwarden-crypto/src/keys/device_key.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+        let decrypted = device_key
+            .decrypt_user_key(protected_device_private_key, protected_user_key)
+            .unwrap();
+
+        assert_eq!(decrypted.0.key, user_key.key,);
+        assert_eq!(decrypted.0.mac_key, user_key.mac_key,);
```
Comment (dani-garcia) [inline crates/bitwarden-crypto/src/keys/device_key.rs:124]:
```suggestion
        assert_eq!(decrypted.0.key, user_key.key);
        assert_eq!(decrypted.0.mac_key, user_key.mac_key);
```

**Review by dani-garcia — DISMISSED** [review]
LGTM. Some minor formatting, not sure why rustfmt didn't catch that

**Review by dani-garcia — APPROVED** [review]

**Pull request merged**